packages/@aws-cdk/assertions/lib/private/mappings.ts

```diff
@@ -0,0 +1,37 @@
+import { Match } from '../match';
+import { Matcher } from '../matcher';
+import { StackInspector } from '../vendored/assert';
+import { formatFailure, matchSection } from './section';
+
+export function findMappings(inspector: StackInspector, props: any = {}): { [key: string]: any }[] {
```

Comment:
Did you consider DRYing up mappings/outputs/resources, or am I over-optimizing too soon? It seems like they could all be written as one method with two extra inputs. I just see a lot of copy/paste between all three files.

```ts
function findSection(inspector: StackInspector, props: any = {}, sectionFn: (x: StackInspector) => any): MatchSuccess | MatchFailure {
  const matcher = Matcher.isMatcher(props) ? props : Match.objectLike(props);
  return matchSection(sectionFn(inspector), matcher);
}

function hasSection(inspector: StackInspector, props: any = {}, sectionFn: (x: StackInspector) => any, sectionName: string): string | void {
  const result = findSection(inspector, props, sectionFn);
  if (result.match) {
    return;
  }

  if (result.closestResult === undefined) {
    return `No ${sectionName} found in the template`;
  }

  return [
    `Template has ${result.analyzedCount} ${sectionName}, but none match as expected.`,
    formatFailure(result.closestResult),
  ].join('\n');
}

export function findMappings(inspector: StackInspector, props: any = {}): { [key: string]: any }[] {
  const result = findSection(inspector, props, (x) => x.value.Mappings ?? {});
  return result.match ? result.matches : [];
}

export function hasMapping(inspector: StackInspector, props: any): string | void {
  return hasSection(inspector, props, (x) => x.value.Mappings ?? {}, 'mappings');
}

export function findOutputs(inspector: StackInspector, props: any = {}): { [key: string]: any }[] {
  const result = findSection(inspector, props, (x) => x.value.Outputs ?? {});
  return result.match ? result.matches : [];
}

export function hasOutput(inspector: StackInspector, props: any): string | void {
  return hasSection(inspector, props, (x) => x.value.Outputs ?? {}, 'outputs');
}
```

(Now that I've seen `section.ts`: Maybe it's just that some/all of this could be moved into matchSection? Certainly the `const matcher = Matcher.isMatcher(props) ? props : Match.objectLike(props);` line could be moved, right?)

Reply:
I have considered this and played around with how much control these methods should retain. It was the usual design choices of 'separation of concerns' and 'layering'. 

I decided to not expose `StackInspector` into the `section` methods and let the individual methods keep control over the values returned / exception messages thrown.

> Certainly the `const matcher = Matcher.isMatcher(props) ? props : Match.objectLike(props);` line could be moved, right?

You are correct about this. Moved this to the layer below.